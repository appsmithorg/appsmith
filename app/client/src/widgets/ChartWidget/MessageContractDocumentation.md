/\*\*

Message protocol between the iframe widget and the main widget

Message Types:

1. click-event
   When a user clicks a datapoint on the chart, iframe sends this to iframe widget wrapper. The event data includes the details of the point clicked by the user.
2. load-complete
   When the iframe has finished loading, it sends this message to the iframe wrapper to let them know. Wrapper then sends the inital chart data to iframe that it needs to render.
3. error
   If there is an exception in setting any chart options, iframe sends this message to the wrapper. It sends the message and stack as string arguments to the wrapper.
4. update-options
   The wrapper sends this message to the iframe to send the options that it would like the iframe to render. It also sends other metata required for rendering those options in the message.

\*/
