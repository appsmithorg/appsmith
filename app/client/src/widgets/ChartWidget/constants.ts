export type ChartType =
  | "LINE_CHART"
  | "BAR_CHART"
  | "PIE_CHART"
  | "COLUMN_CHART"
  | "AREA_CHART"
  | "SCATTER_CHART"
  | "CUSTOM_FUSION_CHART";

export interface ChartDataPoint {
  x: any;
  y: any;
}

export interface ChartData {
  seriesName?: string;
  data: ChartDataPoint[];
}

export interface CustomFusionChartConfig {
  type: string;
  dataSource?: any;
}

export interface AllChartData {
  [key: string]: ChartData;
}

export interface ChartSelectedDataPoint {
  x: any;
  y: any;
  seriesTitle: string;
}

export const CUSTOM_CHART_TYPES = [
  "area2d",
  "bar2d",
  "bar3d",
  "boxandwhisker2d",
  "candlestick",
  "chord",
  "dragnode",
  "dragarea",
  "dragcolumn2d",
  "dragline",
  "errorbar2d",
  "errorline",
  "errorscatter",
  "funnel",
  "gantt",
  "heatmap",
  "hbullet",
  "hled",
  "InverseMSArea",
  "InverseMSColumn2D",
  "InverseMSLine",
  "LogMSColumn2D",
  "LogMSLine",
  "MultiAxisLine",
  "multilevelpie",
  "overlappedcolumn2d",
  "overlappedbar2d",
  "pyramid",
  "radar",
  "angulargauge",
  "realtimearea",
  "bulb",
  "realtimecolumn",
  "cylinder",
  "HLinearGauge",
  "realtimeline",
  "realtimelinedy",
  "realtimestackedarea",
  "realtimestackedcolumn",
  "thermometer",
  "sankey",
  "selectscatter",
  "sparkcolumn",
  "sparkline",
  "sparkwinloss",
  "msstepline",
  "sunburst",
  "treemap",
  "vbullet",
  "vled",
  "waterfall2d",
  "zoomline",
  "zoomlinedy",
  "zoomscatter",
  "column2d",
  "column3d",
  "line",
  "area",
  "bar2d",
  "bar3d",
  "pie2d",
  "pie3d",
  "doughnut2d",
  "doughnut3d",
  "pareto2d",
  "pareto3d",
  "scrollcombidy2d",
  "scrollcombi2d",
  "scrollstackedcolumn2d",
  "scrollmsstackedcolumn2d",
  "scrollmsstackedcolumn2dlinedy",
  "scrollstackedbar2d",
  "scrollarea2d",
  "scrollline2d",
  "scrollcolumn2d",
  "scrollbar2d",
  "bubble",
  "scatter",
  "msstackedcolumn2d",
  "stackedarea2d",
  "stackedbar3d",
  "stackedbar2d",
  "stackedcolumn3d",
  "stackedcolumn2d",
  "msstackedcolumn2dlinedy",
  "stackedcolumn3dlinedy",
  "mscolumn3dlinedy",
  "mscombidy2d",
  "mscombidy3d",
  "stackedcolumn3dline",
  "stackedcolumn2dline",
  "mscolumnline3d",
  "mscombi3d",
  "mscombi2d",
  "marimekko",
  "MSArea",
  "msbar3d",
  "msbar2d",
  "msline",
  "mscolumn3d",
  "mscolumn2d",
  "spline",
  "splinearea",
  "msspline",
  "mssplinedy",
  "mssplinearea",
  "stackedcolumn2dlinedy",
  "stackedarea2dlinedy",
];

export const CUSTOM_CHART_DEFAULT_PARSED = {
  type: "",
  dataSource: {
    chart: {},
    data: [],
  },
};

export enum LabelOrientation {
  AUTO = "auto",
  SLANT = "slant",
  ROTATE = "rotate",
  STAGGER = "stagger",
}

export const LABEL_ORIENTATION_COMPATIBLE_CHARTS = [
  "LINE_CHART",
  "AREA_CHART",
  "COLUMN_CHART",
];
