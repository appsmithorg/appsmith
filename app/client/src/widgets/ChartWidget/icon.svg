<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M7.40002 16.5V22H3V16.5H7.40002Z" fill="#4C5664"/>
<path d="M14 11V22H9.60004V11H14Z" fill="#4C5664"/>
<path d="M20.6002 1.99997V22H16.2002V1.99997H20.6002Z" fill="#4C5664"/>
</svg>
