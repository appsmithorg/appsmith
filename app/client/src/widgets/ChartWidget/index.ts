import { Colors } from "constants/Colors";
import { generateReactKey } from "widgets/WidgetUtils";
import { LabelOrientation } from "./constants";
import IconSVG from "./icon.svg";
import Widget from "./widget";

export const CONFIG = {
  type: Widget.getWidgetType(),
  name: "Chart",
  iconSVG: IconSVG,
  needsMeta: true,
  searchTags: ["graph", "visuals", "visualisations"],
  defaults: {
    rows: 32,
    columns: 24,
    widgetName: "Chart",
    chartType: "COLUMN_CHART",
    chartName: "Sales Report",
    allowScroll: false,
    version: 1,
    animateLoading: true,
    chartData: {
      [generateReactKey()]: {
        seriesName: "Sales",
        data: [
          {
            x: "Product1",
            y: 20000,
          },
          {
            x: "Product2",
            y: 22000,
          },
          {
            x: "Product3",
            y: 32000,
          },
        ],
      },
    },
    xAxisName: "Product Line",
    yAxisName: "Revenue($)",
    labelOrientation: LabelOrientation.AUTO,
    customFusionChartConfig: {
      type: "column2d",
      dataSource: {
        data: [
          {
            label: "Product1",
            value: 20000,
          },
          {
            label: "Product2",
            value: 22000,
          },
          {
            label: "Product3",
            value: 32000,
          },
        ],
        chart: {
          caption: "Sales Report",
          xAxisName: "Product Line",
          yAxisName: "Revenue($)",
          theme: "fusion",
          alignCaptionWithCanvas: 1,
          // Caption styling =======================
          captionFontSize: "24",
          captionAlignment: "center",
          captionPadding: "20",
          captionFontColor: Colors.THUNDER,
          // legend position styling ==========
          legendIconSides: "4",
          legendIconBgAlpha: "100",
          legendIconAlpha: "100",
          legendPosition: "top",
          // Canvas styles ========
          canvasPadding: "0",
          // Chart styling =======
          chartLeftMargin: "20",
          chartTopMargin: "10",
          chartRightMargin: "40",
          chartBottomMargin: "10",
          // Axis name styling ======
          xAxisNameFontSize: "14",
          labelFontSize: "12",
          labelFontColor: Colors.DOVE_GRAY2,
          xAxisNameFontColor: Colors.DOVE_GRAY2,

          yAxisNameFontSize: "14",
          yAxisValueFontSize: "12",
          yAxisValueFontColor: Colors.DOVE_GRAY2,
          yAxisNameFontColor: Colors.DOVE_GRAY2,
        },
      },
    },
  },
  properties: {
    derived: Widget.getDerivedPropertiesMap(),
    default: Widget.getDefaultPropertiesMap(),
    meta: Widget.getMetaPropertiesMap(),
    config: Widget.getPropertyPaneConfig(),
    contentConfig: Widget.getPropertyPaneContentConfig(),
    styleConfig: Widget.getPropertyPaneStyleConfig(),
  },
};

export default Widget;
