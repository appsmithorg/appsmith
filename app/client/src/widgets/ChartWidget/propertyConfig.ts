import { ChartWidgetProps } from "widgets/ChartWidget";

export default [
  {
    sectionName: "General",
    children: [
      {
        helpText: "Adds a title to the chart",
        placeholderText: "Enter title",
        propertyName: "chartName",
        label: "Title",
        controlType: "INPUT_TEXT",
        isBindProperty: true,
        isTriggerProperty: false,
      },
      {
        helpText: "Changes the visualisation of the chart data",
        propertyName: "chartType",
        label: "Chart Type",
        controlType: "DROP_DOWN",
        options: [
          {
            label: "Line Chart",
            value: "LINE_CHART",
          },
          {
            label: "Bar Chart",
            value: "BAR_CHART",
          },
          {
            label: "Pie Chart",
            value: "PIE_CHART",
          },
          {
            label: "Column Chart",
            value: "COLUMN_CHART",
          },
          {
            label: "Area Chart",
            value: "AREA_CHART",
          },
          {
            label: "Custom Chart",
            value: "CUSTOM_FUSION_CHART",
          },
        ],
        isJSConvertible: true,
        isBindProperty: true,
        isTriggerProperty: false,
      },
      {
        propertyName: "isVisible",
        label: "Visible",
        helpText: "Controls the visibility of the widget",
        controlType: "SWITCH",
        isJSConvertible: true,
        isBindProperty: true,
        isTriggerProperty: false,
      },
    ],
  },
  {
    helpText:
      "Manually configure a FusionChart, see https://www.fusioncharts.com",
    propertyName: "customFusionChartConfig",
    placeholderText: `Enter {type: "bar2d","dataSource": {}}`,
    label: "Custom Fusion Chart Configuration",
    controlType: "CUSTOM_FUSION_CHARTS_DATA",
    isBindProperty: true,
    isTriggerProperty: false,
    hidden: (x: any) => x.chartType !== "CUSTOM_FUSION_CHART",
  },
  {
    sectionName: "Chart Data",
    hidden: (props: ChartWidgetProps) =>
      props.chartType === "CUSTOM_FUSION_CHART",
    children: [
      {
        helpText: "Populates the chart with the data",
        propertyName: "chartData",
        placeholderText: 'Enter [{ "x": "val", "y": "val" }]',
        label: "Chart Series",
        controlType: "CHART_DATA",

        isBindProperty: false,
        isTriggerProperty: false,
        children: [
          {
            helpText: "Series Name",
            propertyName: "seriesName",
            label: "Series Name",
            controlType: "INPUT_TEXT",
            isBindProperty: true,
            isTriggerProperty: false,
          },
          {
            helpText: "Series data",
            propertyName: "data",
            label: "Series Data",
            controlType: "INPUT_TEXT_AREA",
            isBindProperty: true,
            isTriggerProperty: false,
          },
        ],
      },
    ],
  },
  {
    sectionName: "Axis",
    hidden: (props: ChartWidgetProps) =>
      props.chartType === "CUSTOM_FUSION_CHART",
    children: [
      {
        helpText: "Specifies the label of the x-axis",
        propertyName: "xAxisName",
        placeholderText: "Enter label text",
        label: "x-axis Label",
        controlType: "INPUT_TEXT",
        isBindProperty: true,
        isTriggerProperty: false,
      },
      {
        helpText: "Specifies the label of the y-axis",
        propertyName: "yAxisName",
        placeholderText: "Enter label text",
        label: "y-axis Label",
        controlType: "INPUT_TEXT",
        isBindProperty: true,
        isTriggerProperty: false,
      },
      {
        helpText: "Enables scrolling inside the chart",
        propertyName: "allowHorizontalScroll",
        label: "Allow horizontal scroll",
        controlType: "SWITCH",
        isBindProperty: false,
        isTriggerProperty: false,
        hidden: (x: any) => x.chartType === "CUSTOM_FUSION_CHART",
      },
    ],
  },
  {
    sectionName: "Actions",
    children: [
      {
        helpText: "Triggers an action when the chart data point is clicked",
        propertyName: "onDataPointClick",
        label: "onDataPointClick",
        controlType: "ACTION_SELECTOR",
        isJSConvertible: true,
        isBindProperty: true,
        isTriggerProperty: true,
      },
    ],
  },
];
