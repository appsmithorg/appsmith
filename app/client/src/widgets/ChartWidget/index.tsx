import React, { lazy, Suspense } from "react";
import BaseWidget, { WidgetProps, WidgetState } from "widgets/BaseWidget";
import { WidgetType } from "constants/WidgetConstants";
import { WidgetPropertyValidationType } from "utils/WidgetValidation";
import { VALIDATION_TYPES } from "constants/WidgetValidation";
import Skeleton from "components/utils/Skeleton";
import * as Sentry from "@sentry/react";
import { retryPromise } from "utils/AppsmithUtils";
import { EventType } from "constants/AppsmithActionConstants/ActionConstants";
import withMeta, { WithMeta } from "widgets/MetaHOC";
import propertyConfig from "widgets/ChartWidget/propertyConfig";
import { CustomFusionChartConfig } from "components/designSystems/appsmith/ChartComponent";

const ChartComponent = lazy(() =>
  retryPromise(() =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "charts" */ "components/designSystems/appsmith/ChartComponent"
    ),
  ),
);

class ChartWidget extends BaseWidget<ChartWidgetProps, WidgetState> {
  static getPropertyValidationMap(): WidgetPropertyValidationType {
    return {
      xAxisName: VALIDATION_TYPES.TEXT,
      yAxisName: VALIDATION_TYPES.TEXT,
      chartName: VALIDATION_TYPES.TEXT,
      isVisible: VALIDATION_TYPES.BOOLEAN,
      chartData: VALIDATION_TYPES.CHART_DATA,
      customFusionChartConfig: VALIDATION_TYPES.CUSTOM_FUSION_CHARTS_DATA,
    };
  }

  static getMetaPropertiesMap(): Record<string, undefined> {
    return {
      selectedDataPoint: undefined,
    };
  }

  static getPropertyPaneConfig() {
    return propertyConfig;
  }

  onDataPointClick = (selectedDataPoint: { x: any; y: any }) => {
    this.props.updateWidgetMetaProperty(
      "selectedDataPoint",
      selectedDataPoint,
      {
        dynamicString: this.props.onDataPointClick,
        event: {
          type: EventType.ON_DATA_POINT_CLICK,
        },
      },
    );
  };

  getPageView() {
    return (
      <Suspense fallback={<Skeleton />}>
        <ChartComponent
          key={this.props.widgetId}
          isVisible={this.props.isVisible}
          chartType={this.props.chartType}
          xAxisName={this.props.xAxisName}
          yAxisName={this.props.yAxisName}
          chartName={this.props.chartName}
          chartData={this.props.chartData}
          customFusionChartConfig={this.props.customFusionChartConfig}
          widgetId={this.props.widgetId}
          onDataPointClick={this.onDataPointClick}
          allowHorizontalScroll={this.props.allowHorizontalScroll}
        />
      </Suspense>
    );
  }

  getWidgetType(): WidgetType {
    return "CHART_WIDGET";
  }
}

export type ChartType =
  | "LINE_CHART"
  | "BAR_CHART"
  | "PIE_CHART"
  | "COLUMN_CHART"
  | "AREA_CHART"
  | "SCATTER_CHART"
  | "CUSTOM_FUSION_CHART";

export interface ChartDataPoint {
  x: any;
  y: any;
}

export interface ChartData {
  seriesName?: string;
  data: ChartDataPoint[];
}

export interface ChartWidgetProps extends WidgetProps, WithMeta {
  chartType: ChartType;
  chartData: ChartData[];
  customFusionChartConfig: { config: CustomFusionChartConfig };
  xAxisName: string;
  yAxisName: string;
  chartName: string;
  isVisible?: boolean;
  allowHorizontalScroll: boolean;
  onDataPointClick?: string;
  selectedDataPoint?: ChartDataPoint;
}

export default ChartWidget;
export const ProfiledChartWidget = Sentry.withProfiler(withMeta(ChartWidget));
