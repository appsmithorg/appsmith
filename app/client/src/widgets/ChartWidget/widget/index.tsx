import React, { lazy, Suspense } from "react";

import type { WidgetProps, WidgetState } from "widgets/BaseWidget";
import BaseWidget from "widgets/BaseWidget";
import Skeleton from "components/utils/Skeleton";
import { retryPromise } from "utils/AppsmithUtils";
import { EventType } from "constants/AppsmithActionConstants/ActionConstants";
import { contentConfig, styleConfig } from "./propertyConfig";
import type {
  ChartType,
  CustomFusionChartConfig,
  AllChartData,
  ChartSelectedDataPoint,
} from "../constants";

import type { WidgetType } from "constants/WidgetConstants";
import type { ChartComponentProps } from "../component";
import { Colors } from "constants/Colors";
import type { Stylesheet } from "entities/AppTheming";
import { DefaultAutocompleteDefinitions } from "widgets/WidgetUtils";
import type { AutocompletionDefinitions } from "widgets/constants";

const ChartComponent = lazy(() =>
  retryPromise(
    () =>
      import(
        /* webpackPrefetch: true, webpackChunkName: "charts" */ "../component"
      ),
  ),
);

class ChartWidget extends BaseWidget<ChartWidgetProps, WidgetState> {
  static getAutocompleteDefinitions(): AutocompletionDefinitions {
    return {
      "!doc":
        "Chart widget is used to view the graphical representation of your data. Chart is the go-to widget for your data visualisation needs.",
      "!url": "https://docs.appsmith.com/widget-reference/chart",
      isVisible: DefaultAutocompleteDefinitions.isVisible,
      chartData: {
        seriesName: "string",
        data: "[$__chartDataPoint__$]",
      },
      xAxisName: "string",
      yAxisName: "string",
      selectedDataPoint: "$__chartDataPoint__$",
    };
  }

  static getMetaPropertiesMap(): Record<string, any> {
    return {
      selectedDataPoint: undefined,
    };
  }

  static getPropertyPaneContentConfig() {
    return contentConfig;
  }

  static getPropertyPaneStyleConfig() {
    return styleConfig;
  }

  static getStylesheetConfig(): Stylesheet {
    return {
      borderRadius: "{{appsmith.theme.borderRadius.appBorderRadius}}",
      boxShadow: "{{appsmith.theme.boxShadow.appBoxShadow}}",
      accentColor: "{{appsmith.theme.colors.primaryColor}}",
      fontFamily: "{{appsmith.theme.fontFamily.appFont}}",
    };
  }

  onDataPointClick = (selectedDataPoint: ChartSelectedDataPoint) => {
    this.props.updateWidgetMetaProperty(
      "selectedDataPoint",
      selectedDataPoint,
      {
        triggerPropertyName: "onDataPointClick",
        dynamicString: this.props.onDataPointClick,
        event: {
          type: EventType.ON_DATA_POINT_CLICK,
        },
      },
    );
  };

  getPageView() {
    return (
      <Suspense fallback={<Skeleton />}>
        <ChartComponent
          allowScroll={this.props.allowScroll}
          borderRadius={this.props.borderRadius}
          boxShadow={this.props.boxShadow}
          chartData={this.props.chartData}
          chartName={this.props.chartName}
          chartType={this.props.chartType}
          customFusionChartConfig={this.props.customFusionChartConfig}
          fontFamily={this.props.fontFamily ?? "Nunito Sans"}
          hasOnDataPointClick={Boolean(this.props.onDataPointClick)}
          isLoading={this.props.isLoading}
          isVisible={this.props.isVisible}
          key={this.props.widgetId}
          labelOrientation={this.props.labelOrientation}
          onDataPointClick={this.onDataPointClick}
          primaryColor={this.props.accentColor ?? Colors.ROYAL_BLUE_2}
          setAdaptiveYMin={this.props.setAdaptiveYMin}
          widgetId={this.props.widgetId}
          xAxisName={this.props.xAxisName}
          yAxisName={this.props.yAxisName}
        />
      </Suspense>
    );
  }

  static getWidgetType(): WidgetType {
    return "CHART_WIDGET";
  }
}
export interface ChartWidgetProps extends WidgetProps {
  chartType: ChartType;
  chartData: AllChartData;
  customFusionChartConfig: CustomFusionChartConfig;
  xAxisName: string;
  yAxisName: string;
  chartName: string;
  isVisible?: boolean;
  allowScroll: boolean;
  borderRadius: string;
  boxShadow?: string;
  accentColor?: string;
  fontFamily?: string;
}

type ChartComponentPartialProps = Omit<ChartComponentProps, "onDataPointClick">;
export interface ChartWidgetProps
  extends WidgetProps,
    ChartComponentPartialProps {
  onDataPointClick?: string;
}

export default ChartWidget;
