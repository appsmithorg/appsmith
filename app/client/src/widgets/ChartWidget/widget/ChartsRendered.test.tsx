import { RenderModes } from "constants/WidgetConstants";
import ChartWidget, { type ChartWidgetProps } from ".";
import { LabelOrientation, type ChartData } from "../constants";

describe("ChartWidget getWidgetView", () => {
  let chartWidget: ChartWidget;

  const seriesData1: ChartData = {
    seriesName: "series1",
    data: [{ x: "x1", y: 1 }],
    color: "series1color",
  };
  const seriesData2: ChartData = {
    seriesName: "series2",
    data: [{ x: "x1", y: 2 }],
    color: "series2color",
  };
  const defaultProps: ChartWidgetProps = {
    allowScroll: true,
    showDataPointLabel: true,
    chartData: {
      seriesID1: seriesData1,
      seriesID2: seriesData2,
    },
    chartName: "chart name",
    type: "CHART_WIDGET",
    chartType: "AREA_CHART",
    customEChartConfig: {},
    customFusionChartConfig: { type: "type", dataSource: undefined },
    hasOnDataPointClick: true,
    isVisible: true,
    isLoading: false,
    setAdaptiveYMin: false,
    labelOrientation: LabelOrientation.AUTO,
    onDataPointClick: "",
    widgetId: "widgetID",
    xAxisName: "xaxisname",
    yAxisName: "yaxisname",
    borderRadius: "1",
    boxShadow: "1",
    primaryColor: "primarycolor",
    fontFamily: "fontfamily",
    dimensions: { componentWidth: 11, componentHeight: 11 },
    parentColumnSpace: 1,
    parentRowSpace: 1,
    topRow: 0,
    bottomRow: 0,
    leftColumn: 0,
    rightColumn: 0,
    widgetName: "widgetName",
    version: 1,
    renderMode: RenderModes.CANVAS,
  };

  it("renders loading state", () => {
    chartWidget = new ChartWidget({ ...defaultProps, isLoading: true });

    const view = chartWidget.getWidgetView();

    expect(view).toMatchSnapshot();
  });

  it("renders error state", () => {
    chartWidget = new ChartWidget({
      ...defaultProps,
      errors: [
        { name: "error", type: "configuration", message: "We have a error" },
      ],
    });

    const view = chartWidget.getWidgetView();

    expect(view).toMatchSnapshot();
  });

  it("renders empty chart data state", () => {
    chartWidget = new ChartWidget({ ...defaultProps, chartData: {} });
    const view = chartWidget.getWidgetView();

    expect(view).toMatchSnapshot();
  });

  it("renders chart with data", () => {
    chartWidget = new ChartWidget(defaultProps);
    const view = chartWidget.getWidgetView();

    expect(view).toMatchSnapshot();
  });
});
