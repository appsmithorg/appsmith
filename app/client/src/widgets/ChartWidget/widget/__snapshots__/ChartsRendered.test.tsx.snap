// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartWidget getWidgetView renders chart with data 1`] = `
<React.Suspense
  fallback={<Skeleton />}
>
  <UNDEFINED
    allowScroll={true}
    borderRadius="1"
    boxShadow="1"
    chartData={
      Object {
        "seriesID1": Object {
          "color": "series1color",
          "data": Array [
            Object {
              "x": "x1",
              "y": 1,
            },
          ],
          "seriesName": "series1",
        },
        "seriesID2": Object {
          "color": "series2color",
          "data": Array [
            Object {
              "x": "x1",
              "y": 2,
            },
          ],
          "seriesName": "series2",
        },
      }
    }
    chartName="chart name"
    chartType="AREA_CHART"
    customEChartConfig={Object {}}
    customFusionChartConfig={
      Object {
        "dataSource": undefined,
        "type": "type",
      }
    }
    dimensions={
      Object {
        "allowScroll": true,
        "borderRadius": "1",
        "bottomRow": 0,
        "boxShadow": "1",
        "chartData": Object {
          "seriesID1": Object {
            "color": "series1color",
            "data": Array [
              Object {
                "x": "x1",
                "y": 1,
              },
            ],
            "seriesName": "series1",
          },
          "seriesID2": Object {
            "color": "series2color",
            "data": Array [
              Object {
                "x": "x1",
                "y": 2,
              },
            ],
            "seriesName": "series2",
          },
        },
        "chartName": "chart name",
        "chartType": "AREA_CHART",
        "customEChartConfig": Object {},
        "customFusionChartConfig": Object {
          "dataSource": undefined,
          "type": "type",
        },
        "dimensions": Object {
          "componentHeight": 11,
          "componentWidth": 11,
        },
        "fontFamily": "fontfamily",
        "hasOnDataPointClick": true,
        "isLoading": false,
        "isVisible": true,
        "labelOrientation": "auto",
        "leftColumn": 0,
        "onDataPointClick": "",
        "parentColumnSpace": 1,
        "parentRowSpace": 1,
        "primaryColor": "primarycolor",
        "renderMode": "CANVAS",
        "rightColumn": 0,
        "setAdaptiveYMin": false,
        "showDataPointLabel": true,
        "topRow": 0,
        "type": "CHART_WIDGET",
        "version": 1,
        "widgetId": "widgetID",
        "widgetName": "widgetName",
        "xAxisName": "xaxisname",
        "yAxisName": "yaxisname",
      }
    }
    fontFamily="Nunito Sans"
    hasOnDataPointClick={false}
    isLoading={false}
    isVisible={true}
    labelOrientation="auto"
    onDataPointClick={[Function]}
    primaryColor="#553DE9"
    setAdaptiveYMin={false}
    showDataPointLabel={true}
    widgetId="widgetID"
    xAxisName="xaxisname"
    yAxisName="yaxisname"
  />
</React.Suspense>
`;

exports[`ChartWidget getWidgetView renders empty chart data state 1`] = `<EmptyChartData />`;

exports[`ChartWidget getWidgetView renders error state 1`] = `
<ChartErrorComponent
  error={
    Object {
      "message": "We have a error",
      "name": "error",
      "type": "configuration",
    }
  }
/>
`;

exports[`ChartWidget getWidgetView renders loading state 1`] = `
<React.Suspense
  fallback={<Skeleton />}
>
  <UNDEFINED
    allowScroll={true}
    borderRadius="1"
    boxShadow="1"
    chartData={
      Object {
        "seriesID1": Object {
          "color": "series1color",
          "data": Array [
            Object {
              "x": "x1",
              "y": 1,
            },
          ],
          "seriesName": "series1",
        },
        "seriesID2": Object {
          "color": "series2color",
          "data": Array [
            Object {
              "x": "x1",
              "y": 2,
            },
          ],
          "seriesName": "series2",
        },
      }
    }
    chartName="chart name"
    chartType="AREA_CHART"
    customEChartConfig={Object {}}
    customFusionChartConfig={
      Object {
        "dataSource": undefined,
        "type": "type",
      }
    }
    dimensions={
      Object {
        "allowScroll": true,
        "borderRadius": "1",
        "bottomRow": 0,
        "boxShadow": "1",
        "chartData": Object {
          "seriesID1": Object {
            "color": "series1color",
            "data": Array [
              Object {
                "x": "x1",
                "y": 1,
              },
            ],
            "seriesName": "series1",
          },
          "seriesID2": Object {
            "color": "series2color",
            "data": Array [
              Object {
                "x": "x1",
                "y": 2,
              },
            ],
            "seriesName": "series2",
          },
        },
        "chartName": "chart name",
        "chartType": "AREA_CHART",
        "customEChartConfig": Object {},
        "customFusionChartConfig": Object {
          "dataSource": undefined,
          "type": "type",
        },
        "dimensions": Object {
          "componentHeight": 11,
          "componentWidth": 11,
        },
        "fontFamily": "fontfamily",
        "hasOnDataPointClick": true,
        "isLoading": true,
        "isVisible": true,
        "labelOrientation": "auto",
        "leftColumn": 0,
        "onDataPointClick": "",
        "parentColumnSpace": 1,
        "parentRowSpace": 1,
        "primaryColor": "primarycolor",
        "renderMode": "CANVAS",
        "rightColumn": 0,
        "setAdaptiveYMin": false,
        "showDataPointLabel": true,
        "topRow": 0,
        "type": "CHART_WIDGET",
        "version": 1,
        "widgetId": "widgetID",
        "widgetName": "widgetName",
        "xAxisName": "xaxisname",
        "yAxisName": "yaxisname",
      }
    }
    fontFamily="Nunito Sans"
    hasOnDataPointClick={false}
    isLoading={true}
    isVisible={true}
    labelOrientation="auto"
    onDataPointClick={[Function]}
    primaryColor="#553DE9"
    setAdaptiveYMin={false}
    showDataPointLabel={true}
    widgetId="widgetID"
    xAxisName="xaxisname"
    yAxisName="yaxisname"
  />
</React.Suspense>
`;
